### DISCUSSION TITLE
Require explicit path when using `import`

### DISCUSSION THREAD

**Discussion by jimmywarting** [babel/babel#15017] [Ideas]
importing some file without an `.js` extension while using ESM syntax is a bad practices.
extensionless is a byproduct from NodeJS that just so happens to work b/c ppl are still transpiling ESM things to CJS and using `require()` to resolve the paths form you. either that or you are packing everything into one file. 

If you would not use any transpiler and just write plain vanilla ESM then you would know that when you are using NodeJS with `"type": "module"`, using remote http-import or simply just use `<script type="module">` then you would have to use explicit paths b/c it wouldn't work otherwise. b/c you don't have access to the filesystem and isn't able to just simply scan the file system of what files you need. 

For you to still allow this behavior to work then you are enabling developers to write "bad" code.
It's bad b/c what if somebody desire to just import some parts of someone else uncompiled library src, then it won't work b/c they don't use explicit path. 
Someone who have realized this problem is for instance twitter bootstrap (in https://github.com/twbs/bootstrap/issues/31944)
they now regret that they didn't use explicit path from the beginning. 

Also if you have +5000 of files and everyone of this require some other file without an extension then that will be a lots of stating and scanning for the correct file. if you are more explicit then NodeJS don't have to guess as much and loading files becomes some few sec faster to load (https://nolanlawson.com/2016/08/15/the-cost-of-small-modules/)

---

I guess what i'm asking for is: if you scan some code that looks like this:
```js
import xyz from './foo'
```
then dump a error or a warning message that says the file `./foo` could not be found. or using inexplicit path won't work in the feature or something like that.

**Comment by nicolo-ribaudo** [comment]
That's valid JavaScript, so Babel is not going to throw an error these. It doesn't work in Node.js (unless you use a custom resolver) but it can work, for example, in browsers.

This is a linter's job, which can disallow bad practices. Babel only cares about per-file correctness.

**Reply by jimmywarting** [reply]
But when you walk the dependency tree then and goes into `/foo.js` then i assume that babel have used `require.resolve` to resolve some modules? i rather wish that it used `import.meta.resolve` instead...

**Reply by jimmywarting** [reply]
it's not only a linters job. it's also your fault to allow this bad behavior to still work the way that it dose. 
if someone is writing stuff with ESM and using babel then i think they would have to opt in to using the old legacy `require.resolve` instead if they want to enable the bad extensionless behavior

**Reply by nicolo-ribaudo** [reply]
> But when you walk the dependency tree

Babel works on a per-file basis, it does not walk the dependency tree. Bundlers do that.

**Reply by jimmywarting** [reply]
ah... my fault then.
